source/platforms/allwinner-d1/src/main.rs

```diff
@@ -0,0 +1,623 @@
+#![no_std]
+#![no_main]
+
+use core::cell::UnsafeCell;
+use core::ptr::{NonNull, null_mut};
+use core::sync::atomic::{compiler_fence, Ordering, fence, AtomicPtr, AtomicU64};
+
+use kernel::comms::bbq::{new_bidi_channel, SpscProducer};
+use kernel::comms::kchannel::{KConsumer, KChannel};
+use kernel::comms::oneshot::Reusable;
+use kernel::registry::simple_serial::{SimpleSerial, Request, Response, SimpleSerialError};
+use kernel::registry::{RegisteredDriver, KernelHandle, ReplyTo, Envelope};
+use kernel::{self, Kernel, registry::Message};
+
+use d1_pac::{Interrupt, TIMER, UART0, DMAC};
+use d1_playground::dmac::descriptor::{
+    AddressMode, BModeSel, BlockSize, DataWidth, DescriptorConfig, DestDrqType, SrcDrqType, Descriptor,
+};
+use d1_playground::dmac::{Dmac, ChannelMode, Channel};
+use maitake::sync::Mutex;
+use maitake::wait::{WaitCell, WaitQueue, Closed};
+use mnemos_alloc::containers::{HeapArc, HeapArray};
+
+use d1_playground::plic::{Plic, Priority};
+use d1_playground::timer::{Timer, TimerMode, TimerPrescaler, TimerSource, Timers};
+
+use uuid::{Uuid, uuid};
+
+struct Uart(d1_pac::UART0);
+static mut PRINTER: Option<Uart> = None;
+impl core::fmt::Write for Uart {
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        while self.0.usr.read().tfnf().bit_is_clear() {}
+        for byte in s.as_bytes() {
+            self.0.thr().write(|w| unsafe { w.thr().bits(*byte) });
+            while self.0.usr.read().tfnf().bit_is_clear() {}
+        }
+        Ok(())
+    }
+}
+fn print_raw(data: &[u8]) {
+    let uart = unsafe { PRINTER.as_mut().unwrap() };
+    while uart.0.usr.read().tfnf().bit_is_clear() {}
+    for byte in data {
+        uart.0.thr().write(|w| unsafe { w.thr().bits(*byte) });
+        while uart.0.usr.read().tfnf().bit_is_clear() {}
+    }
+}
+pub fn _print(args: core::fmt::Arguments) {
+    use core::fmt::Write;
+    unsafe {
+        PRINTER.as_mut().unwrap().write_fmt(args).ok();
+    }
+}
+#[macro_export]
+macro_rules! print {
+    ($($arg:tt)*) => {
+        $crate::_print(core::format_args!($($arg)*));
+    }
+}
+#[macro_export]
+macro_rules! println {
+    ($($arg:tt)*) => {
+        $crate::_print(core::format_args!($($arg)*));
+        $crate::print!("\r\n");
+    }
+}
+
+static TICK_MS: AtomicU64 = AtomicU64::new(0);
+static TICK_WAKER: AtomicPtr<WaitCell> = AtomicPtr::new(null_mut());
+static UART_TX_WAKER: AtomicPtr<WaitCell> = AtomicPtr::new(null_mut());
```

Comment:
FWIW, `WaitCell::new` is a const fn, so I _think_ that rather than allocating `HeapArc`s and storing them in an `AtomicPtr`, you could just have a
```suggestion
static TICK_WAKER: WaitCell = WaitCell::new();
static UART_TX_WAKER: WaitCell: WaitCell::new();
```

(of course, this would mean modifying other code as well)

Reply:
Oh, duh, good call!